tensorboard/data/server/server.rs

```diff
@@ -44,12 +60,33 @@ impl TensorBoardDataProvider for DataProviderHandler {
         &self,
         _request: Request<data::ListRunsRequest>,
     ) -> Result<Response<data::ListRunsResponse>, Status> {
-        let mut res: data::ListRunsResponse = Default::default();
-        res.runs.push(data::Run {
-            name: "train".to_string(),
-            start_time: FAKE_START_TIME,
+        let runs = self.read_runs()?;
+
+        // Buffer up started runs to sort by wall time. Keep `WallTime` rather than projecting down
+        // to f64 so that we're guaranteed that they're non-NaN and can sort them.
+        let mut results: Vec<(Run, WallTime)> = Vec::with_capacity(runs.len());
+        for (run, data) in runs.iter() {
+            let data = data
+                .read()
+                .map_err(|_| Status::internal(format!("failed to read run data for {:?}", run)))?;
+            if let Some(start_time) = data.start_time {
+                results.push((run.clone(), start_time));
+            }
+        }
+        results.sort_by_key(|&(_, start_time)| start_time);
+        drop(runs); // release lock a bit earlier
+
+        let res = data::ListRunsResponse {
+            runs: results
+                .into_iter()
+                .map(|(Run(name), start_time)| data::Run {
+                    name,
+                    start_time: start_time.into(),
+                    ..Default::default()
```

Comment:
Yeah—specifically, so that proto updates that add new fields don’t
introduce compilation errors.

I was considering adding a tiny “style guide” to `DEVELOPMENT.md` with
things like this, so I’ll probably go ahead and do that.

There is a `#[non_exhaustive]` attribute that you can add to structs to
indicate that they may have more fields later, preventing callers
outside the crate from initializing them directly but still permitting
them to access public fields. This doesn’t suffice because (a) we’re in
the same crate (though we could change that) and (b) it doesn’t permit
functional record updates (`..Default::default()`):

```
error[E0639]: cannot create non-exhaustive struct using struct expression
 --> main.rs:3:13
  |
3 |       let s = sub::S {
  |  _____________^
4 | |         a: 1,
5 | |         b: 2,
6 | |         ..Default::default()
7 | |     };
  | |_____^
```

The RFC threads have a lot of comments, and from skimming them it’s not
obvious to me whether this was “intended, by design” or merely “artifact
of how `#[non_exhaustive]` desugars to a hidden private field”. (It was
definitely _known_ at the time.)

Another potential check for this would be in Clippy: “lint if you
construct a value at a type that implements `prost::Message` without
using a functional record update”. There is no such check currently.

Reply:
Reading through the next PR, it's really rather noisy to have the `..Default::default()` everywhere we create a proto, especially in tests :/  Given the fact that we control these protos ourselves, I wonder if on net maybe it's easier for places where we don't expect to add fields to just relax this and accept that we'll need to make modifications to the code if we update the proto?

The other alternative that came to mind would be to initialize structs to default-empty, then mutate to populate fields; needing mutation is a downside although I think in some of the more convoluted multi-layer struct cases this may be significantly more concise.

I can see why neither of these is ideal, it just feels distracting IMO to have every single proto struct need the extra syntax.  Would be nice if there was some way to specify on the struct definition that they can be initialized a subset of fields if the rest are `Default` (basically like optional arguments), or if at least maybe the functional update syntax supported plain `..` as sugar for `..Default::default()`.

Reply:
Just seeing this. For posterity, further discussion at #4399.